File: src/main/java/com/pubnub/api/models/server/access_manager/v3/GrantTokenRequestBody.java
package com.pubnub.api.models.server.access_manager.v3;

import com.google.gson.JsonObject;
import com.pubnub.api.PubNub;
import com.pubnub.api.PubNubException;
import com.pubnub.api.builder.PubNubErrorBuilder;
import com.pubnub.api.models.consumer.access_manager.v3.Channel;
import com.pubnub.api.models.consumer.access_manager.v3.Group;
import com.pubnub.api.models.consumer.access_manager.v3.PNResource;
import com.pubnub.api.models.consumer.access_manager.v3.Space;
import com.pubnub.api.models.consumer.access_manager.v3.User;
import lombok.Builder;
import lombok.extern.java.Log;

import java.util.List;

@Log
@Builder
public class GrantTokenRequestBody {

    private static final int READ = 1;
    private static final int WRITE = 2;
    private static final int MANAGE = 4;
    private static final int DELETE = 8;
    private static final int CREATE = 16;

    private Integer ttl;
    private List<Channel> channels;
    private List<Group> groups;
    private List<User> users;
    private List<Space> spaces;
    private Object meta;
    private PubNub pubNub;

    public JsonObject assemble() throws PubNubException {
        JsonObject payload = new JsonObject();

        payload.addProperty("ttl", this.ttl);

        JsonObject permissions = new JsonObject();

        JsonObject resources = new JsonObject();
        JsonObject patterns = new JsonObject();

        parse(channels, "channels", resources, patterns);
        parse(groups, "groups", resources, patterns);
        parse(users, "users", resources, patterns);
        parse(spaces, "spaces", resources, patterns);

        permissions.add("resources", resources);
        permissions.add("patterns", patterns);

        if (this.meta != null) {
            try {
                permissions.add("meta", pubNub.getMapper().convertValue(this.meta, JsonObject.class));
            } catch (PubNubException e) {
                throw PubNubException.builder()
                        .pubnubError(PubNubErrorBuilder.PNERROBJ_INVALID_META)
                        .build();
            }
        } else {
            permissions.add("meta", new JsonObject());
        }

        payload.add("permissions", permissions);

        return payload;
    }

    private void parse(List<? extends PNResource> list, String resourceSetName, JsonObject resources,
                       JsonObject patterns) throws PubNubException {
        if (list != null) {
            for (PNResource pnResource : list) {
                JsonObject resourceObject = new JsonObject();

                JsonObject determinedObject;

                if (pnResource.isPatternResource()) {
                    determinedObject = patterns;
                } else {
                    determinedObject = resources;
                }

                if (determinedObject.has(resourceSetName)) {
                    determinedObject.get(resourceSetName).getAsJsonObject()
                            .addProperty(pnResource.getId(), calculateBitmask(pnResource));
                } else {
                    resourceObject.addProperty(pnResource.getId(), calculateBitmask(pnResource));
                    determinedObject.add(resourceSetName, resourceObject);
                }
            }
        }

        if (!resources.has(resourceSetName)) {
            resources.add(resourceSetName, new JsonObject());
        }
        if (!patterns.has(resourceSetName)) {
            patterns.add(resourceSetName, new JsonObject());
        }
    }

    private int calculateBitmask(PNResource resource) throws PubNubException {
        int sum = 0;
        if (resource.isRead()) {
            sum += READ;
        }
        if (resource.isWrite()) {
            sum += WRITE;
        }
        if (resource.isManage()) {
            sum += MANAGE;
        }
        if (resource.isDelete()) {
            sum += DELETE;
        }
        if (resource.isCreate()) {
            sum += CREATE;
        }
        if (sum == 0) {
            throw PubNubException.builder()
                    .pubnubError(PubNubErrorBuilder.PNERROBJ_PERMISSION_MISSING)
                    .errormsg("No permissions specified for resource: ".concat(resource.getId()))
                    .build();
        }
        return sum;
    }
}

File: src/main/kotlin/com/pubnub/api/models/server/access_manager/v3/GrantTokenRequestBody.kt
package com.pubnub.api.models.server.access_manager.v3

import com.pubnub.api.PubNubException
import com.pubnub.api.models.TokenBitmask
import com.pubnub.api.models.consumer.access_manager.v3.ChannelGrant
import com.pubnub.api.models.consumer.access_manager.v3.ChannelGroupGrant
import com.pubnub.api.models.consumer.access_manager.v3.PNGrant
import com.pubnub.api.models.consumer.access_manager.v3.PNPatternGrant
import com.pubnub.api.models.consumer.access_manager.v3.UUIDGrant

data class GrantTokenRequestBody(
    val ttl: Int,
    val permissions: GrantTokenPermissions
) {

    data class GrantTokenPermissions(
        val resources: GrantTokenPermission,
        val patterns: GrantTokenPermission,
        val meta: Any? = null,
        val uuid: String? = null
    )

    data class GrantTokenPermission(
        val channels: Map<String, Int> = emptyMap(),
        val groups: Map<String, Int> = emptyMap(),
        val uuids: Map<String, Int> = emptyMap(),
        val spaces: Map<String, Int> = emptyMap(),
        val users: Map<String, Int> = emptyMap()
    )

    companion object {
        @Throws(PubNubException::class)
        fun of(
            ttl: Int,
            channels: List<ChannelGrant>,
            groups: List<ChannelGroupGrant>,
            uuids: List<UUIDGrant>,
            meta: Any?,
            uuid: String?
        ): GrantTokenRequestBody {
            val resources = GrantTokenPermission(
                getResources(channels),
                getResources(groups), getResources(uuids)
            )
            val patterns = GrantTokenPermission(
                getPatterns(channels),
                getPatterns(groups), getPatterns(uuids)
            )
            val permissions = GrantTokenPermissions(resources, patterns, meta ?: emptyMap<Any, Any>(), uuid)
            return GrantTokenRequestBody(ttl, permissions)
        }

        private fun <T : PNGrant> getResources(resources: List<T>): Map<String, Int> {
            return resources
                .filter { it !is PNPatternGrant }
                .associate { it.id to calculateBitmask(it) }
        }

        private fun <T : PNGrant> getPatterns(resources: List<T>): Map<String, Int> {
            return resources
                .filterIsInstance(PNPatternGrant::class.java)
                .associate { it.id to calculateBitmask(it) }
        }

        private fun calculateBitmask(resource: PNGrant): Int {
            var sum = 0
            if (resource.read) {
                sum = sum or TokenBitmask.READ
            }
            if (resource.write) {
                sum = sum or TokenBitmask.WRITE
            }
            if (resource.manage) {
                sum = sum or TokenBitmask.MANAGE
            }
            if (resource.delete) {
                sum = sum or TokenBitmask.DELETE
            }
            if (resource.create) {
                sum = sum or TokenBitmask.CREATE
            }
            if (resource.get) {
                sum = sum or TokenBitmask.GET
            }
            if (resource.join) {
                sum = sum or TokenBitmask.JOIN
            }
            if (resource.update) {
                sum = sum or TokenBitmask.UPDATE
            }
            return sum
        }
    }
}
